### PR TITLE
[sagas] Make a macro to simplify declaring saga actions

### PR DESCRIPTION
**Pull request by smklein**
Writing sagas involves a lot of boilerplate:

- When creating `NexusAction` nodes, it's convention to re-state the name of the saga in each label (and also name the label after the variable itself). This is easily machine-generated.
- After defining all the `NexusAction` nodes, they must then all be re-listed in `register_actions`.
- When actually creating the DAG for the saga, each node is re-stated. This is the spot where node "output names" can be defined, but the label is another variation of the `NexusAction` node name (usually camelcased).

This PR introduces (and uses) a macro called `declare_saga_actions`.

This macro aims to reduce this boilerplate, by requiring only the following:
- The name of the saga
- The name of each action
- The output of each action
- The "forward" action function
- (Optional) The "undo" action function

It then generates nodes with names that are automatically "the right format", and produces some helper functions:
- `fn <saga name>_register_actions(registry: &mut ActionRegistry)` to help implement `NexusSaga::register_actions`
- `fn <node>_action() -> Node` to produce a default `Node` implementation when generating the DAG

### CODE REVIEW

**Comment by zephraph** [comment]
This is so much better! I'm gonna leave it to other folks to review because I don't have faith in my macro foo, but thanks for doing this.

File: nexus/src/app/sagas/mod.rs
Comment (sunshowers) [inline nexus/src/app/sagas/mod.rs:210]: Can we use `once_cell::Lazy` here (maybe in a followup since this just refactors code?) `lazy_static` is mostly deprecated at this point I think.

File: nexus/src/app/sagas/snapshot_create.rs
Comment (sunshowers) [inline nexus/src/app/sagas/snapshot_create.rs:164]: Being less error-prone here is really nice!

File: nexus/src/app/sagas/snapshot_create.rs
Comment (sunshowers) [inline nexus/src/app/sagas/snapshot_create.rs:195]: Can this use the same pattern as `register_actions` above where you don't have to specify all the actions?

File: nexus/src/app/sagas/mod.rs
Comment (sunshowers) [inline nexus/src/app/sagas/mod.rs:182]: Do we expect to make this a public macro (`macro_export`) in the future, maybe for other parts of the code that use sagas?

**Review by sunshowers — APPROVED** [review]
I think this macro does a good job enhancing the clarity of the code. Just a few questions but nothing blocking.

**Review by leftwo — APPROVED** [review]
Love it.

Would it be a good idea to put a block comment somewhere to help new saga authors
figure out what they need to provide to the macro, much like the write up you have
here for explaining the PR itself?   Somethign like this list:

```
This macro aims to reduce this boilerplate, by requiring only the following:

The name of the saga
The name of each action
The output of each action
The "forward" action function
(Optional) The "undo" action function
```

File: nexus/src/app/sagas/mod.rs
Comment (smklein) [reply nexus/src/app/sagas/mod.rs]: Done in 6abce0865dfb225b3581f05df7a544b2561168f8 ! We are using `lazy_static` elsewhere in Omicron, but that can be converted too

File: nexus/src/app/sagas/snapshot_create.rs
Comment (smklein) [reply nexus/src/app/sagas/snapshot_create.rs:195]: Ah, good question. This is a part I didn't feel comfortable automating yet.

In the case of *this particular saga* - `snapshot_create` - we just call all the actions that we declared, in order.

However, not all our sagas are this simple - the DAG in `make_saga_dag` describes the control flow of execution, which may be more complex than "all the actions we declared in the order of declaration".

For example...

... Constants can be embedded into the DAG:

https://github.com/oxidecomputer/omicron/blob/2f453125d0ecb95f091706976368589c073a7d35/nexus/src/app/sagas/instance_create.rs#L142-L147

... Nodes that are shared among multiple sagas can be called (such as generating UUIDs):

https://github.com/oxidecomputer/omicron/blob/2f453125d0ecb95f091706976368589c073a7d35/nexus/src/app/sagas/instance_create.rs#L149-L153

... Some nodes can be called in parallel, or called as subsagas:

https://github.com/oxidecomputer/omicron/blob/2f453125d0ecb95f091706976368589c073a7d35/nexus/src/app/sagas/instance_create.rs#L223-L245

File: nexus/src/app/sagas/mod.rs
Comment (smklein) [reply nexus/src/app/sagas/mod.rs:188]: Good question! Right now, I expected to keep this "Nexus-internal", since my short-term goal was to just make it easier to add new sagas.

However, if we like the structure of this macro, I could see us making it public, or moving it into the `steno` repo itself.

**Comment by smklein** [comment]
> Love it.
> 
> Would it be a good idea to put a block comment somewhere to help new saga authors figure out what they need to provide to the macro, much like the write up you have here for explaining the PR itself? Somethign like this list:
> 
> ```
> This macro aims to reduce this boilerplate, by requiring only the following:
> 
> The name of the saga
> The name of each action
> The output of each action
> The "forward" action function
> (Optional) The "undo" action function
> ```

I extended the macro's documentation to list this within https://github.com/oxidecomputer/omicron/pull/2058/commits/29a3ab5a00001777ef2a1a738c81a549051fb37b

**Comment by davepacheco** [comment]
Thanks for doing this.  I'm sorry I didn't get to the review before it landed.  It's really nice to have a lot less repetition here.

One thing to keep in mind for the future is that there are reasons the action name, node label, and output name might be different.  The only example I found today was with the common uuid.generate action (which is kind of obviated after dynamic DAGs anyway).  The idea is you might have a common piece of code (the action) that's used in multiple places in the graph or multiple different graphs to generate different outputs (e.g., instance_id, disk_id, etc.).  I don't think this changes much here.  I think it just means that if someone decides it would be useful to have different action name / node label / node name, they can't use this macro or they need to generalize the macro first somehow.  (It's also possible that allowing these to be different isn't that useful an idea and Steno shouldn't require so many strings that in practice are going to be all the same.)

**Pull request merged**
